ompi/mca/coll/basic/coll_basic_neighbor_allgatherv.c

```diff
@@ -64,36 +63,40 @@ mca_coll_basic_neighbor_allgatherv_cart(const void *sbuf, int scount, struct omp
         }
 
         if (MPI_PROC_NULL != srank) {
+            nreqs += 2;
             rc = MCA_PML_CALL(irecv((char *) rbuf + disps[i] * extent, rcounts[i], rdtype, srank,
-                                    MCA_COLL_BASE_TAG_ALLGATHER, comm, reqs++));
+                                    MCA_COLL_BASE_TAG_ALLGATHER, comm, preqs++));
             if (OMPI_SUCCESS != rc) break;
 
             /* remove cast from const when the pml layer is updated to take
              * a const for the send buffer. */
             rc = MCA_PML_CALL(isend((void *) sbuf, scount, sdtype, srank, MCA_COLL_BASE_TAG_ALLGATHER,
-                                    MCA_PML_BASE_SEND_STANDARD, comm, reqs++));
+                                    MCA_PML_BASE_SEND_STANDARD, comm, preqs++));
             if (OMPI_SUCCESS != rc) break;
-            nreqs += 2;
         }
 
         if (MPI_PROC_NULL != drank) {
+            nreqs += 2;
```

Comment:
@bosilca Replace nreqs+=2 by nreqs++ before each request. That way, we'll be able to
free axactly the right number of requests in case of failure.